compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1338,14 +1395,15 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                             }),
                         ))
                     }
-                    ImplTraitContext::Disallowed(_) => {
+                    ImplTraitContext::Disallowed(position) => {
                         let mut err = struct_span_err!(
                             self.sess,
                             t.span,
                             E0562,
-                            "`impl Trait` not allowed outside of {}",
-                            "function and method return types",
+                            "`impl Trait` not allowed outside of \
+                            function and inherent method return types",
                         );
+                        err.note(&format!("found `impl Trait` in {}", position));
```

Comment:
I'm concerned about "inherent" in this wording not being too clear. "associated functions" doesn't communicate the same meaning, sadly. Either way, I would likely change this to be 

```suggestion
                            "`impl Trait` only allowed in function and inherent method return types, not in {}", position
                        );
```